File: internal/scaffold/scaffold.go
// Package scaffold implements `iBuild init`: it materializes a new OKF-SDLC
// bundle from embedded templates. The base type profile is copied verbatim as
// DATA — the taxonomy is never encoded in Go. Init never overwrites an existing
// file, so it is safe to re-run and safe in a partially set-up repo.
package scaffold

import (
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// Options tune a scaffold run.
type Options struct {
	// Example also writes a tiny, validate-clean example requirement.
	Example bool
	// Full selects the full SDLC taxonomy instead of the lean core profile.
	Full bool
}

// Result reports what a scaffold run did. Paths are bundle-relative, slash-separated.
type Result struct {
	Created     []string
	Skipped     []string
	AlreadyInit bool // .ibuildos.yaml already existed at the target
}

const embedRoot = "templates"

// profilesPrefix holds the swappable type profiles (core/ and full/). The main
// copy skips it; the chosen profile is copied into docs/types/ separately.
const profilesPrefix = embedRoot + "/profiles/"

// example* are minimal, internally-consistent requirements with only the inherited
// required fields, at `proposed` (a proposed, unimplemented requirement is a
// warning, never an error), so the init→validate round trip stays green with
// --example. The type must match the scaffolded profile.
const exampleReqPath = "docs/requirements/req-0001.md"
const exampleReqCore = `---
type: Requirement
id: REQ-0001
title: Example requirement — replace or delete me
owner: you
status: proposed
---

Scaffolded by ` + "`iBuild init --example`" + `. Replace it with your real
requirements, then delete this file.
`

const exampleFRPath = "docs/requirements/fr-0001.md"
const exampleFRFull = `---
type: FunctionalRequirement
id: FR-0001
title: Example requirement — replace or delete me
owner: you
status: proposed
---

Scaffolded by ` + "`iBuild init --example --full`" + `. Use /ibuild-discover and
/ibuild-plan to build out the real chain, then delete this file.
`

// Init scaffolds target into an iBuildOS bundle and returns what it created vs.
// skipped. It never overwrites existing files.
func Init(target string, opts Options) (Result, error) {
	var res Result
	if _, err := os.Stat(filepath.Join(target, ".ibuildos.yaml")); err == nil {
		res.AlreadyInit = true
	}

	// Copy the shared bundle skeleton, skipping the swappable type profiles.
	err := fs.WalkDir(templatesFS, embedRoot, func(p string, d fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if d.IsDir() || strings.HasPrefix(p, profilesPrefix) {
			return nil
		}
		rel := strings.TrimPrefix(p, embedRoot+"/")
		if err := writeFile(target, destPath(rel), mustRead(p), &res); err != nil {
			return err
		}
		return nil
	})
	if err != nil {
		return res, err
	}

	// Copy the chosen type profile (core by default) into docs/types/.
	profile := "core"
	if opts.Full {
		profile = "full"
	}
	entries, err := fs.ReadDir(templatesFS, profilesPrefix+profile)
	if err != nil {
		return res, err
	}
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		src := profilesPrefix + profile + "/" + e.Name()
		if err := writeFile(target, "docs/types/"+e.Name(), mustRead(src), &res); err != nil {
			return res, err
		}
	}

	if opts.Example {
		path, body := exampleReqPath, exampleReqCore
		if opts.Full {
			path, body = exampleFRPath, exampleFRFull
		}
		if err := writeFile(target, path, []byte(body), &res); err != nil {
			return res, err
		}
	}

	sort.Strings(res.Created)
	sort.Strings(res.Skipped)
	return res, nil
}

// destPath maps an embedded template path to its on-disk destination. The only
// special case is the dot-stripped config name (go:embed can store but some
// tooling dislikes a dot-prefixed source file, so it is stored as ibuildos.yaml).
func destPath(rel string) string {
	if rel == "ibuildos.yaml" {
		return ".ibuildos.yaml"
	}
	return rel
}

// writeFile writes data to <target>/<dest>, creating parents, unless dest already
// exists (in which case it is recorded as skipped). dest is slash-separated.
func writeFile(target, dest string, data []byte, res *Result) error {
	full := filepath.Join(target, filepath.FromSlash(dest))
	if _, err := os.Stat(full); err == nil {
		res.Skipped = append(res.Skipped, dest)
		return nil
	} else if !os.IsNotExist(err) {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(full), 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(full, data, 0o644); err != nil {
		return err
	}
	res.Created = append(res.Created, dest)
	return nil
}

func mustRead(p string) []byte {
	b, err := templatesFS.ReadFile(p)
	if err != nil {
		// Unreachable: p comes from WalkDir over the same FS.
		panic(err)
	}
	return b
}

File: src/core/scaffold/scaffold.ts
// `iBuild init`: materializes a new OKF-SDLC bundle from the embedded templates.
// The base type profile is copied verbatim as DATA — the taxonomy is never
// encoded in code. Init never overwrites an existing file, so it is safe to
// re-run. Port of Go internal/scaffold.
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { join, dirname } from "node:path";
import { TEMPLATES } from "./embedded.ts";

export interface InitOptions {
  example?: boolean; // also write a tiny, validate-clean example requirement
  full?: boolean; // select the full SDLC taxonomy instead of the lean core profile
}

export interface InitResult {
  created: string[]; // bundle-relative, slash-separated
  skipped: string[];
  alreadyInit: boolean; // .ibuildos.yaml already existed at the target
}

const PROFILES_PREFIX = "profiles/";

// Minimal, internally-consistent example requirements at `proposed` (a proposed,
// unimplemented requirement is a warning, never an error) so init→validate stays
// green with --example. The type must match the scaffolded profile.
const EXAMPLE_REQ_PATH = "docs/requirements/req-0001.md";
const EXAMPLE_REQ_CORE = `---
type: Requirement
id: REQ-0001
title: Example requirement — replace or delete me
owner: you
status: proposed
---

Scaffolded by \`iBuild init --example\`. Replace it with your real requirements,
then delete this file.
`;
const EXAMPLE_FR_PATH = "docs/requirements/fr-0001.md";
const EXAMPLE_FR_FULL = `---
type: FunctionalRequirement
id: FR-0001
title: Example requirement — replace or delete me
owner: you
status: proposed
---

Scaffolded by \`iBuild init --example --full\`. Use /ibuild-discover and
/ibuild-plan to build out the real chain, then delete this file.
`;

function writeFile(target: string, dest: string, data: string, res: InitResult): void {
  const full = join(target, dest);
  if (existsSync(full)) {
    res.skipped.push(dest);
    return;
  }
  mkdirSync(dirname(full), { recursive: true });
  writeFileSync(full, data);
  res.created.push(dest);
}

// init scaffolds target into an iBuildOS bundle. It never overwrites existing files.
export function init(target: string, opts: InitOptions = {}): InitResult {
  const res: InitResult = { created: [], skipped: [], alreadyInit: existsSync(join(target, ".ibuildos.yaml")) };

  // Shared bundle skeleton, skipping the swappable type profiles.
  for (const key of Object.keys(TEMPLATES).sort()) {
    if (key.startsWith(PROFILES_PREFIX)) continue;
    const dest = key === "ibuildos.yaml" ? ".ibuildos.yaml" : key;
    writeFile(target, dest, TEMPLATES[key]!, res);
  }

  // Chosen type profile (core by default) into docs/types/.
  const profile = opts.full ? "full" : "core";
  const pfx = `${PROFILES_PREFIX}${profile}/`;
  for (const key of Object.keys(TEMPLATES).sort()) {
    if (!key.startsWith(pfx)) continue;
    writeFile(target, `docs/types/${key.slice(pfx.length)}`, TEMPLATES[key]!, res);
  }

  if (opts.example) {
    const path = opts.full ? EXAMPLE_FR_PATH : EXAMPLE_REQ_PATH;
    const body = opts.full ? EXAMPLE_FR_FULL : EXAMPLE_REQ_CORE;
    writeFile(target, path, body, res);
  }

  res.created.sort();
  res.skipped.sort();
  return res;
}
